runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/ActiveWorkRefresher.java

```diff
@@ -115,32 +132,72 @@ private void invalidateStuckCommits() {
     }
   }
 
+  /** Create {@link Heartbeats} and group them by {@link HeartbeatSender}. */
   private void refreshActiveWork() {
     Instant refreshDeadline = clock.get().minus(Duration.millis(activeWorkRefreshPeriodMillis));
+    Map<HeartbeatSender, Heartbeats> heartbeatsBySender =
+        aggregateHeartbeatsBySender(refreshDeadline);
 
+    if (heartbeatsBySender.isEmpty()) {
+      return;
+    }
+
+    if (heartbeatsBySender.size() == 1) {
+      // If there is a single HeartbeatSender, just use the calling thread to send heartbeats.
+      Map.Entry<HeartbeatSender, Heartbeats> heartbeat =
+          Iterables.getOnlyElement(heartbeatsBySender.entrySet());
+      sendHeartbeat(heartbeat);
+    } else {
+      // If there are multiple HeartbeatSenders, send out the heartbeats in parallel using the
+      // fanOutActiveWorkRefreshExecutor.
+      List<CompletableFuture<Void>> fanOutRefreshActiveWork = new ArrayList<>();
+      for (Map.Entry<HeartbeatSender, Heartbeats> heartbeat : heartbeatsBySender.entrySet()) {
+        fanOutRefreshActiveWork.add(
+            CompletableFuture.runAsync(
+                () -> sendHeartbeat(heartbeat), fanOutActiveWorkRefreshExecutor));
+      }
+
+      // Don't block until we kick off all the refresh active work RPCs.
```

Comment:
It would be equivalent to just loop over fanOutRefreshActiveWork and join each individually (if we're not surfacing exceptions from them it behaves the same and seems likely less code).

Reply:
Could also combine inlining single element to reduce threads used in fanout case by 1 doing something like:

// Send one heartbeat using this thread and start rest async.
@Nullable Map.Entry<HeartbeatSender, Heartbeats> inlineHeartbeat = null;
 List<CompletableFuture<Void>> fanOutRefreshActiveWork = new ArrayList<>();
      for (Map.Entry<HeartbeatSender, Heartbeats> heartbeat : heartbeatsBySender.entrySet()) {
  if (inlineHeartbeat == null) {
    inlineHeartbeat = heartbeat;
  } else {
    // runAsync
  }
}
sendHeartbeat(inlineHeartbeat);
// loop and block on async

Reply:
done

is it a large benefit if we reduce the fanout by 1?